Fix default value for boolean type in SQLite getting columns schema

batopa opened the pull request:
This PR fixes an issue with default value of boolean type columns for SQLite database.

Building a table with `TableSchema` for example:

```php
$s = new TableSchema('examples');
$s->addColumn('field1', [
    'type' => 'boolean',
    'null' => true,
    'default' => true,
]);
```
will produce

```SQL
CREATE TABLE "examples" ("field1" BOOLEAN DEFAULT TRUE)
```
Once the table was created reading the schema we obtain an inconsistency columns for default value. 

```php
$schema = new SchemaCollection($connection);
$result = $schema->describe('examples');
debug($result->getColumn('field1'));
```

output

```
########## DEBUG ##########
[
  'type' => 'boolean',
  'length' => null,
  'null' => true,
  'default' => 'TRUE',
  'comment' => null,
  'precision' => null
]
###########################
```

The default is a string `'TRUE'` but I expect a boolean `true` instead.

The same happens defining default as INT, for example `0`. The results in this case will be `'default' => '0'`.

---

**Side note**

To change code just in one place I modified the `protected function _defaultValue` signature. I think that it shouldn't have serious impact but let me know in case.

batopa committed:
Fix default value getting columns schema in sqlite

dereuromark commented on tests/TestCase/Database/Schema/SqliteSchemaDialectTest.php:
```diff
@@ -293,6 +293,7 @@ protected function _createTables($connection): void
 author_id INT(11) NOT NULL,
 unique_id UNSIGNED INTEGER NOT NULL,
 published BOOLEAN DEFAULT 0,
+reviewed BOOLEAN DEFAULT TRUE,
```
Why do we need this? Wouldnt it suffice to just map the other ones to numeric `0` as well?

dereuromark replied:
Or maybe we just switch to using TRUE/FALSE then explicitly (in the future)?

batopa replied:
:thinking: I think that it is a more complete test case because `\Cake\Database\Schema\SqliteSchemaDialect::columnSql()`:

* passing `'default' => true` produces` BOOLEAN DEFAULT TRUE`
* passing `'default' => false` produces `BOOLEAN DEFAULT FALSE` 
* passing `'default' => 1` produces `BOOLEAN DEFAULT 1`
* passing `'default' => 0` produces `BOOLEAN DEFAULT 0` 

so you can deal with both cases. 

Anyway if you think it is unnecessary I can remove it.

dereuromark replied:
Based on the other code changes above I think it should for now always produce only 0 and 1.

markstory replied:
For reference, SQLite doesn't have a dedicated boolean type

>SQLite does not have a separate Boolean storage class. Instead, Boolean values are stored as integers 0 (false) and 1 (true).
>
>SQLite recognizes the keywords "TRUE" and "FALSE", as of version 3.23.0 (2018-04-02) but those keywords are really just alternative spellings for the integer literals 1 and 0 respectively.

https://sqlite.org/datatype3.html

Our other drivers use `0`/`1`  in their schema reflection boolean logic, and it would be good to be consistent here.

dereuromark commented on src/Database/Schema/SqliteSchemaDialect.php:
```diff
@@ -304,14 +304,19 @@ public function describeColumns(string $tableName): array
             return null;
         }
 
+        if (strtolower($type) === TableSchemaInterface::TYPE_BOOLEAN) {
```
```suggestion
        if ($type !== null && strtolower($type) === TableSchemaInterface::TYPE_BOOLEAN) {
```

dereuromark commented:
I guess the one fail is probably an invalid test (data):
 ```
1) Cake\Test\TestCase\Database\Schema\SqliteSchemaDialectTest::testConvertColumn with data set #3 ('BOOLEAN', ['boolean', null])
Failed asserting that two arrays are identical.
--- Expected
+++ Actual
@@ @@
 Array &0 [
     'comment' => null,
-    'default' => 'Default value',
+    'default' => false,
     'length' => null,
     'null' => true,
     'type' => 'boolean',
 ]
 ```
 Since it is "not empty" I wonder if this should map to `true` instead?

dereuromark commented on src/Database/Schema/SqliteSchemaDialect.php:
```diff
@@ -304,14 +304,19 @@ public function describeColumns(string $tableName): array
+     * @return string|int|bool|null
      */
-    protected function _defaultValue(string|int|null $default): string|int|null
+    protected function _defaultValue(string|int|null $default, ?string $type = null): string|int|bool|null
```
How about as bugfix for 5.x:
- You dont change the signature and map it to 0/1

Then in 5.next:
- We can adjust the internal protected method to always return true/false for booleans from now on.

This is a bit more BC safe for people and consistent in the output then.

batopa replied:
Do you mean to leave the second parameter `?string $type = null` and change the return value for type boolean always to `int`?

dereuromark replied:
Minimal changeset that makes the code work, yes, for a bugfix release.

markstory replied:
I don't think we need to worry too much about BC here. I doubt that there are userland extensions to the sqlite schema dialect.

dereuromark replied:
I still think it would be cleaner if we had one single way of interpretation.

> Our other drivers use 0/1 in their schema reflection boolean logic, and it would be good to be consistent here.

So its settled then.

batopa committed:
fix as suggested

Co-authored-by: Mark Scherer <dereuromark@users.noreply.github.com>

batopa commented:
> I guess the one fail is probably an invalid test (data):
> 
> ```
> 1) Cake\Test\TestCase\Database\Schema\SqliteSchemaDialectTest::testConvertColumn with data set #3 ('BOOLEAN', ['boolean', null])
> Failed asserting that two arrays are identical.
> --- Expected
> +++ Actual
> @@ @@
> Array &0 [
>     'comment' => null,
> -    'default' => 'Default value',
> +    'default' => false,
>     'length' => null,
>     'null' => true,
>     'type' => 'boolean',
> ]
> ```
> 
> Since it is "not empty" I wonder if this should map to `true` instead?

Uhm... I'm not sure how to handle this situation.
The issue is that `filter_var($default, FILTER_VALIDATE_BOOLEAN)` return `false` trying to validate `'Default value'`.
We could add the flag [FILTER_NULL_ON_FAILURE](https://www.php.net/manual/it/filter.constants.php#constant.filter-null-on-failure) to trace that the default is not valid and take a decision on top of that.

But what is the action to take in this scenario? SQLite allows to set any default i.e. `CREATE table examples (field BOOLEAN DEFAULT not_a_boolean);` won't throw any error.

Any suggestion?

dereuromark commented:
> But what is the action to take in this scenario? SQLite allows to set any default i.e. CREATE table examples (field BOOLEAN DEFAULT not_a_boolean); won't throw any error.

What does it convert it to? 0 or 1? whatever the real result, thats your answer IMO.

markstory committed:
Add my suggestions to 18973

- Fixed failing tests
- Aligned on int cast for boolean column default values.
- Made validation stricter by not handling arbitrary string-integers

markstory commented:
I pushed up a commit with my suggestions and fixed failing tests.

ADmad commented on src/Database/Schema/SqliteSchemaDialect.php:
```diff
@@ -304,14 +304,23 @@ public function describeColumns(string $tableName): array
      * @param string|int|null $default The default value.
-     * @return string|int|null
+     * @param string|null $type The column type.
+     * @return string|int|bool|null
```
```suggestion
     * @return string|int|null
```

batopa committed:
update @return signature

Co-authored-by: ADmad <ADmad@users.noreply.github.com>
